src/kafka/protocol/messages.js

```diff
@@ -202,6 +239,89 @@ const parseCommand = (messageProtocol) => {
   }
 }
 
+/**
+ * Encodes Payload to base64 encoded data URI
+ *
+ * @param {buffer\|string} input - Buffer or String
+ * @param {MimeTypes} mimeType - mime type of the input
+ *
+ * @return {string} - Returns base64 encoded data  URI string
+ */
+
+exports.encodePayload = (input, mimeType) => {
+  if (allowedMimeTypes.includes(mimeType)) {
+    return (input instanceof Buffer)
+      ? `data:${mimeType};base64,${base64url(input, 'utf8')}`
+      : `data:${mimeType};base64,${base64url(Buffer.from(input), 'utf8')}`
+  } else {
+    throw new Error(`mime type should be one of the following:${allowedMimeTypes.map(e => ` ${e}`)}`)
+  }
+}
+
+/**
+ * Decode Payload to base64 encoded data URI
+ *
+ * @param {string} input - JSON as Data URI or plain string
+ * @param {object} [options = {asParsed: true}] - Parising object
+ *
+ * @return {(object\|Protocol~DecodedURI)} based on the options, returns parsed JSON or decodedURI object
+ */
+
+const decodePayload = (input, { asParsed = true } = {}) => {
+  const dataUriRegEx = /^\s*data:'?(?:([\w-]+)\/([\w+.-]+))'??(?:;charset=([\w-]+))?(?:;(base64))?,(.*)/
+
+  const parseDecodedDataToJson = (decodedData) => {
+    if (['application/json', 'application/vnd.interoperability.transfers+json;version=1'].includes(decodedData.mimeType)) return JSON.parse(decodedData.data.toString())
+    else if (decodedData.mimeType === 'text/plain') return decodedData.data.toString()
+    else throw new Error('invalid mime type')
+  }
+
+  try {
+    if (RegExp(dataUriRegEx).test(input)) {
+      let decoded = parseDataUri(input)
+      return asParsed
+        ? parseDecodedDataToJson(decoded)
+        : decoded
+    } else if (typeof input === 'string') {
+      return asParsed ? JSON.parse(input) : { mimeType: 'text/plain', data: input }
+    } else {
+      throw new Error('input should be Buffer or String')
+    }
+  } catch (e) {
+    throw e
+  }
+}
+
+/**
+ * Decode message or messages
+ *
+ * @param {(Protocol~Message\|Protocol~Message[])} messages - single message or array of messages with payload encoded as base64 dataURI
+ * @param {object} [options = {asParsed: true}] - options to parse the payload or not
+ *
+ * @returns {(Protocol~Message\|Protocol~Message[])} - messages with decoded payload
+ */
+
+exports.decodeMessages = (messages) => {
+  const decodeMessage = (message) => {
+    let payload = decodePayload(message.value.content.payload)
+    message.value.content.payload = payload
```

Comment:
Note that the message should be copied/deep-clined, and then have the content re-assigned. Aka we don't want to modify the original message here.